chore: Verify CN name as part of client cert check while authenticating api server (#2396)
open-policy-agent/gatekeeper: website/docs/externaldata.md

### Authenticate the API server against Webhook (Self managed K8s cluster only)

> ⚠️ Two new flags will be introduced in v3.11 because of these changes. And since they are not backward compatible, you may need a clean install to make use of them.

**Note:** To enable authenticating the API server you have to be able to modify cluster resources. This may not be possible for managed K8s clusters.

To ensure a request to the Gatekeeper webhook is coming from the API server, Gatekeeper needs to validate the client cert in the request. To enable authenticate API server, the following configuration can be made:

> 1. Deploy Gatekeeper with a client CA cert name. Provide name of the client CA with the flag `--client-ca-name`. The same name will be used to read certificate from the webhook secret. The webhook will only authenticate API server requests if client CA name is provided with flag.

Reviewer: Might be good to provide an example of how this is configured as part of the deployment yaml and/or helm chart. might be good to update https://github.com/open-policy-agent/gatekeeper-external-data-provider to reflect this once v3.11 is out.

1. Deploy Gatekeeper with a client CA cert name. Provide name of the client CA with the flag `--client-cert-name`. The same name will be used to read certificate from the webhook secret. The webhook will only authenticate API server requests if client CA name is provided with flag.

> 2. You will need to patch the webhook secret manually to attach client ca crt. Update secret `gatekeeper-webhook-server-cert` to include `clientca.crt`. Key name `clientca.crt` should match the name passed with `--client-cert-name` flag. For client CA you could either use default kubernetes general ca located at `\etc\kubernetes\pki\ca.crt`, or you could generate your own CA for this purpose.

Reviewer 2: ```suggestion
2. You will need to patch the webhook secret manually to attach client ca crt. Update secret `gatekeeper-webhook-server-cert` to include `clientca.crt`. Key name `clientca.crt` should match the name passed with `--client-cert-name` flag. For client CA you could either use the default Kubernetes CA located at `\etc\kubernetes\pki\ca.crt`, or you could generate your own CA for this purpose.
```

Reviewer 3: We need to stop recommending that people mix trust domains.

Author: Got it!

      ca.key: <ca-key>
      tls.crt: <tls-crt>
      tls.key: <tls-key>
      clientca.crt: <ca-crt>
    metadata:
      ...
      name: <gatekeeper-webhook-service-name>
      namespace: <gatekeeper-namespace>
    type: Opaque
    ```

> 3. Generate client certificates signed with CA authority to be attached by API server while talking to Gatekeeper webhook. Gatekeeper webhook expects API server to attach the certificate that has CN name as `kube-apiserver`. If certificate with anyother CN name is used, webhook will throw the error and not accept the request.

Reviewer 2: ```suggestion
3. Generate client certificates signed with CA authority to be attached by the API server while talking to Gatekeeper webhook. Gatekeeper webhook expects the API server to attach the certificate that has CN name as `kube-apiserver`. If a certificate with any other CN name is used, webhook will throw the error and not accept the request.
```

    - Generate private key for API server

      ```bash
      openssl genrsa -out apiserver-client.key 2048
      ```
    - Generate a CSR for API server
    
      ```bash
      openssl req -new -key apiserver-client.key -out apiserver-client.csr -subj="/CN=kube-apiserver"
      ```
    
    - Generate public key for API server
      | Replace `<ca-crt>` and `<ca-key>` with correct values

      ```bash
      openssl x509 -req -in apiserver-client.csr -CA <ca-crt> -CAkey <ca-key> -CAcreateserial -out apiserver-client.crt -days 825
      ```

> 4. You will need to make sure the K8s API Server includes appropriate certificate while sending requests to the webhook, otherwise webhook will not accept these requests and will log error of `tls client didn't provide a certificate`. To make sure API server attaches correct certificate to requests being sent to webhook, you must specify the location of the admission control configuration file via the `--admission-control-config-file` flag while starting the API server. Here is an example admission control configuration file:

Reviewer 2: ```suggestion
4. You will need to make sure the K8s API Server includes the appropriate certificate while sending requests to the webhook, otherwise webhook will not accept these requests and will log an error of `tls client didn't provide a certificate`. To make sure the API server attaches the correct certificate to requests being sent to webhook, you must specify the location of the admission control configuration file via the `--admission-control-config-file` flag while starting the API server. Here is an example admission control configuration file:
```

    apiVersion: v1
    clusters:
    - cluster:
        certificate-authority-data: <ca-crt> # same value as provided in gatekeeper webhook secret's clientca.crt 
        server: https://<gatekeeper-webhook-service-name>.<gatekeeper-namespace>.svc:443
      name: <cluster-name>
    contexts:
    - context:
        cluster: <cluster-name>
        user: api-server
      name: <context-name>
    current-context: <context-name>
    kind: Config
    users:
    - name: api-server
      user:

>         client-certificate-data: <apiserver-client.crt> 

Reviewer: nit: you are using `apiserver-client.crt` here but `ca-crt` and `clientca.crt` else where. perhaps make them consistent to reduce confusion.

Author: Done! updated the doc to make it consistent.
